components/containerd/pod/component.go

```diff
@@ -100,95 +149,170 @@ func (c *component) Metrics(ctx context.Context, since time.Time) ([]components.
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+func createConn(ctx context.Context, endpoint string) (runtimeapi.RuntimeServiceClient, runtimeapi.ImageServiceClient, *grpc.ClientConn, error) {
+	// "k8s.io/cri-client/pkg/util.GetAddressAndDialer" doesn't work...
+	// "code = Unavailable desc = name resolver error: produced zero addresses"
+	addr, err := parseUnixEndpoint(endpoint)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// "WithBlock" ctx cancel is no-op
+	conn, err := grpc.DialContext(ctx, addr, defaultDialOptions()...) //nolint:staticcheck
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	log.Logger.Debugw("successfully dialed -- checking version", "endpoint", endpoint)
+
+	// ref. https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+	runtimeClient := runtimeapi.NewRuntimeServiceClient(conn)
+	version, err := runtimeClient.Version(ctx, &runtimeapi.VersionRequest{})
+	if err != nil {
+		conn.Close()
+		return nil, nil, nil, err
 	}
+	log.Logger.Debugw("successfully checked version", "endpoint", endpoint, "version", version.String())
 
-	b, _ := o.JSON()
-	return []components.State{{
-		Name:    containerd_pod_id.Name,
-		Health:  components.StateHealthy,
-		Healthy: true,
-		Reason:  o.describeReason(),
-		ExtraInfo: map[string]string{
-			"data":     string(b),
-			"encoding": "json",
-		},
-	}}, nil
+	status, err := runtimeClient.Status(ctx, &runtimeapi.StatusRequest{})
+	if err != nil {
+		conn.Close()
+		return nil, nil, nil, err
+	}
+	log.Logger.Debugw("successfully checked status", "endpoint", endpoint, "status", status.String())
+
+	imageClient := runtimeapi.NewImageServiceClient(conn)
+	return runtimeClient, imageClient, conn, nil
 }
 
-const (
-	defaultSocketFile               = "/run/containerd/containerd.sock"
-	defaultContainerRuntimeEndpoint = "unix:///run/containerd/containerd.sock"
-)
+func CheckContainerdRunning(ctx context.Context) bool {
+	if runtime.GOOS != "linux" {
+		log.Logger.Debugw("ignoring default containerd pod checking since it's not linux", "os", runtime.GOOS)
+		return false
+	}
+
+	p, err := pkg_file.LocateExecutable("containerd")
+	if err == nil {
+		log.Logger.Debugw("containerd found in PATH", "path", p)
+		return true
+	}
+	log.Logger.Debugw("containerd not found in PATH -- fallback to containerd run checks", "error", err)
+
+	containerdSocketExists := false
+	containerdRunning := false
 
-func listSandboxStatus(ctx context.Context, endpoint string) ([]*runtimeapi.PodSandboxStatusResponse, error) {
+	if _, err := os.Stat(defaultSocketFile); err == nil {
+		log.Logger.Debugw("containerd default socket file exists, containerd installed", "file", defaultSocketFile)
+		containerdSocketExists = true
+	} else {
+		log.Logger.Debugw("containerd default socket file does not exist, skip containerd check", "file", defaultSocketFile, "error", err)
+	}
```

Comment:
IMO we should check the unix socket first, then check binary.. because we rely on the socket.
And I'm not sure that if we stopped containerd service, does the socket file exist?